src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -1238,6 +1249,34 @@ internal override void AddSynthesizedAttributes(PEModuleBuilder moduleBuilder, r
             }
         }
 
+        /// <summary>
+        /// Returns true if the type is generic or non-generic custom task-like type due to the
+        /// [AsyncMethodBuilder(typeof(B))] attribute. It returns the "B".
+        /// </summary>
+        /// <param name="builderArgument">The argument to the <see cref="System.Runtime.CompilerServices.AsyncMethodBuilderAttribute"/> or null if not a custom task</param>
+        /// <returns>True if this type is a custom task type</returns>
+        /// <remarks>
+        /// The definition of "custom task-like type" is one that has an [AsyncMethodBuilder(typeof(B))] attribute,
+        /// no more, no less. Validation of builder type B is left for elsewhere. This method returns B
+        /// without validation of any kind.
+        /// 
+        /// This specialization retrieves B from the early decoded attributed data: this prevents binding loops 
+        /// when we're binding attributes on a type and need to know during binding if the same type is task-like
+        /// </remarks>
+        public override bool IsCustomTaskType(out object builderArgument)
+        {
+            var earlyDecodeData = this.GetEarlyDecodedWellKnownAttributeData();
+            if (earlyDecodeData != null
+                && earlyDecodeData.AsyncMethodBuilderTarget.Kind == TypedConstantKind.Type)
+            {
+                builderArgument = earlyDecodeData.AsyncMethodBuilderTarget.Value;
+                return true;
```

Comment:
> return true; [](start = 15, length = 13)

Can we get false positive from this method? Can it return true when base implementation would return false? For example, when constructor takes object instead of System.Type?